src/cares_wrap.cc

```diff
@@ -296,6 +300,124 @@ Local<Array> HostentToNames(Environment* env, struct hostent* host) {
   return scope.Escape(names);
 }
 
+void safe_free_hostent(struct hostent* host) {
+  int idx;
+
+  if (host->h_addr_list) {
+    idx = 0;
+    while (host->h_addr_list[idx]) {
+      free(host->h_addr_list[idx++]);
+    }
+    free(host->h_addr_list);
+    host->h_addr_list = 0;
+  }
+
+  if (host->h_aliases) {
+    idx = 0;
+    while (host->h_aliases[idx]) {
+      free(host->h_aliases[idx++]);
+    }
+    free(host->h_aliases);
+    host->h_aliases = 0;
+  }
+
+  if (host->h_name) {
+    free(host->h_name);
+  }
+
+  host->h_addrtype = host->h_length = 0;
+}
+
+bool cares_wrap_hostent_cpy(struct hostent* dest, struct hostent* src) {
+  dest->h_addr_list = nullptr;
+  dest->h_addrtype = 0;
+  dest->h_aliases = nullptr;
+  dest->h_length = 0;
+  dest->h_name = nullptr;
+
+  /* copy `h_name` */
+  unsigned int name_size = (strlen(src->h_name) + 1) * sizeof(char);
+  dest->h_name = reinterpret_cast<char*>(node::Malloc(name_size));
+  if (!dest->h_name) return false;
+  memcpy(dest->h_name, src->h_name, name_size);
+
+  /* copy `h_aliases` */
+  unsigned int alias_count;
+  unsigned int alias_size;
+  unsigned int cur_alias_length;
+  for (alias_count = 0;
+      src->h_aliases[alias_count] != nullptr;
+      alias_count++) {
+  }
+  alias_size = sizeof(char*) * (alias_count + 1);
+
+  dest->h_aliases = reinterpret_cast<char**>(node::Malloc(alias_size));
+  if (dest->h_aliases == nullptr) {
+    safe_free_hostent(dest);
+    return false;
+  }
+
+  memset(reinterpret_cast<void*>(dest->h_aliases), 0, alias_size);
+  for (unsigned int i = 0; i < alias_count; i++) {
+    cur_alias_length = strlen(src->h_aliases[i]);
+    dest->h_aliases[i] = reinterpret_cast<char*>(
+        node::Malloc(sizeof(char) * (cur_alias_length + 1)));
+
+    if (dest->h_aliases[i] == nullptr) {
```

Comment:
`node::Malloc()` aborts if it sees an OOM – I personally like code that does checking, but you’ll want `node::UncheckedMalloc()` for the variant that can return a `nullptr`.